### ISSUE TITLE
Splatter parameter estimation error.

### ISSUE THREAD

**Issue by AlemuTA** [Oshlack/splatter#31]
Hello

I was trying to use splatter to simulate scRNA-seq data starting from a source dataset (real scRNA-seq data), but I encountered the following message (error) when I try to estimate the parameters using splatEstimate() function.  It is related to fitting distributions. I tried it with subset of the data and also with two other data  but I couldn't resolve the issue. Can you check it out?

Thanks,
Alemu

-------------------------------------------------------------------------------------
<simpleError in optim(par = vstart, fn = fnobj, fix.arg = fix.arg, obs = data,     gr = gradient, pdistnam = pdistname, hessian = TRUE, method = meth,     lower = lower, upper = upper, ...): non-finite finite-difference value [2]>
Error in nls(y ~ logistic(x, x0 = x0, k = k), data = df, start = list(x0 = 0,  : 
  step factor 0.000488281 reduced below 'minFactor' of 0.000976562
In addition: Warning message:
In splatEstMean(norm.counts, params) :
  Goodness of fit failed, using Method of Moments
----------------------------------------------------------------------------------------------------

**Comment by lazappi** [comment]
Hi @AlemuTA

Thanks for giving Splatter a go. I hope it can be useful for you once we get this sorted out.

Can you please let me know what version of Splatter, R and Bioconductor you are using?

It looks like there is something about the data you are using that is causing the dropout fitting process to fail, possibly this is just because there isn't any dropout in your data. Can you tell me anything more about the dataset you are using (or point me to it if it is public)? Please make sure you are using raw counts as input to the estimation function, rather than FPKM or some other normalised values.

Luke

**Comment by AlemuTA** [comment]
Hi Luke

Thanks for your response. First, I appreciate your great work. Your simulation approach is comprehensive, flexible, and  easy to implement. 

I am using version "1.2.1" of splatter package,  R 3.4.3, and the latest Bioconductor.

I used raw counts of a publicly available scRNA-seq data (ReCount database accession number SRP060416), and you can access the processed data (.rds file) here https://www.dropbox.com/s/3ekayu1e1ue2zby/proccessed.tonsil.data.rds?dl=0  and the codes I used to process the raw data (.R file) is here https://www.dropbox.com/s/44i5io0jn8b96um/prepprocessing%20tonsile%20data.R?dl=0, and the original downloaded data can be found here https://www.dropbox.com/s/lq7vrwxvxbmv9jk/rse_gene.Rdata?dl=0 

So far, I estimated the gamma parameters (for mean expression), the log-normal parameters (for library size), and the BCV from edgeR  manually and carried on simulating scRNA-data.  However, still other tuning parameters need to be a reflect of my source data, and I hope you can figure it out.

Regards,
Alemu

**Comment by lazappi** [comment]
Hi Alemu

Thanks for providing the data and code, that really helped. I tried doing the estimation and got the same message, but that is actually ok, it is still working (I'll explain more below). Here is what I did:

```{r}
> sce <- readRDS("~/Downloads/proccessed.tonsil.data.rds")
> params <- splatEstimate(sce)
<simpleError in optim(par = vstart, fn = fnobj, fix.arg = fix.arg, obs = data,     gr = gradient, pdistnam = pdistname, hessian = TRUE, method = meth,     lower = lower, upper = upper, ...): non-finite finite-difference value [2]>
Warning message:
In splatEstMean(norm.counts, params) :
  Goodness of fit failed, using Method of Moments
> params
A Params object of class SplatParams 
Parameters can be (estimable) or [not estimable], 'Default' or 'NOT DEFAULT'. 

Global: 
(GENES)  (CELLS)   [Seed]  
  35992      568   692396  

27 additional parameters 

Batches: 
    [Batches]  [BATCH CELLS]     [Location]        [Scale]  
            1            568            0.1            0.1  

Mean: 
              (RATE)               (SHAPE)  
0.000910036817657866     0.393620292254069  

Library size: 
       (LOCATION)            (SCALE)  
  17.376724408629  0.654393256646933  

Exprs outliers: 
       (PROBABILITY)            (LOCATION)               (SCALE)  
0.000500180620779726      8.89734072946663     0.827540888319642  

Groups: 
     [Groups]  [Group Probs]  
            1              1  

Diff expr: 
[Probability]    [Down Prob]     [Location]        [Scale]  
          0.1            0.5            0.1            0.4  

BCV: 
   (COMMON DISP)             (DOF)  
9.06422837114371  12.7273708375139  

Dropout: 
         [Present]          (MIDPOINT)             (SHAPE)  
             FALSE    8.35791361476919  -0.493428863651076  

Paths: 
        [From]        [Length]          [Skew]    [Non-linear]  [Sigma Factor]  
             0             100             0.5             0.1             0.8  

> sim <- splatSimulate(params)
Getting parameters...
Creating simulation object...
Simulating library sizes...
Simulating gene means...
Simulating BCV...
Simulating counts..
Simulating dropout (if needed)...
Done!
> sim
class: SingleCellExperiment 
dim: 35992 568 
metadata(1): params
assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
rownames(35992): Gene1 Gene2 ... Gene35991 Gene35992
rowData names(4): Gene BaseGeneMean OutlierFactor GeneMean
colnames(568): Cell1 Cell2 ... Cell567 Cell568
colData names(3): Cell Batch ExpLibSize
reducedDimNames(0):
spikeNames(0):
```

Hopefully you can see here that despite getting the message Splatter has still successfully estimated the parameters and those parameters can be used to simulate a new dataset.

One of the estimation steps is fitting a gamma distribution to the mean expression of the genes. There are different methods you can use to do this fitting. When we were developing Splatter we found that the Goodness of Fit method gave the best estimates. Unfortunately this method fails on some datasets (which is the error message you are seeing). When this happens we automatically switch to the Method of Moments method, which gives slightly worse results but is more robust (this is what the warning message is trying to say). Someone who knows more about stats can probably explain why this is. 

In short Splatter is working ok and you should be safe to ignore the messages and use the parameters it returns. I will update the warning message in future versions to better explain what is going on.

Hope that helps

Luke

**Comment by lazappi** [comment]
I'm going to close this issue but feel free to comment if you have more questions.

**Comment by AlemuTA** [comment]
Dear Luke,

I am sorry for not to replying fast.

I really appreciate your suggestion. I am going to ignore the warning as you suggested.

Cheers.
Alemu
________________________________
